File: 9_排序和搜索/经典题/自定义比较函数/937. 重新排列日志文件.py
# 字母日志：除标识符之外，所有字均由小写字母组成
# 数字日志：除标识符之外，所有字均由数字组成

#  请按下述规则将日志重新排序：

#  所有 字母日志 都排在 数字日志 之前。
#  字母日志 在内容不同时，忽略标识符后，按内容字母顺序排序；在内容相同时，按标识符排序。
#  数字日志 应该保留原来的相对顺序。


# reorderLogFiles(['dig1 8 1 5 1', 'let1 art can', 'dig2 3 6', 'let2 own kit dig', 'let3 art zero'])

#  输出：["let1 art can","let3 art zero","let2 own kit dig","dig1 8 1 5 1","dig2 3 6"]
#  解释：
#  字母日志的内容都不同，所以顺序为 "art can", "art zero", "own kit dig" 。
#  数字日志保留原来的相对顺序 "dig1 8 1 5 1", "dig2 3 6" 。
from typing import Any, List, Tuple


class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        def compare_key(log: str) -> Tuple[Any, ...]:
            left, right = log.split(" ", 1)
            if right[0].isalpha():
                return (0, right, left)
            else:
                return (1,)

        return sorted(logs, key=compare_key)

File: 9_排序和搜索/经典题/自定义比较函数/937. 重新排列日志文件.ts
// 所有 字母日志 都排在 数字日志 之前。
// 字母日志 在内容不同时，忽略标识符后，按内容字母顺序排序；在内容相同时，按标识符排序。
function reorderLogFiles(logs: string[]): string[] {
  const comparator = (a: string, b: string): number => {
    const [[mark1, ...content1], [mark2, ...content2]] = [a.split(' '), b.split(' ')]
    const [isDigit1, isDigit2] = [
      Number.isInteger(Number(content1[0])),
      Number.isInteger(Number(content2[0])),
    ]

    return isDigit1 && isDigit2
      ? 0
      : Number(isDigit1) - Number(isDigit2) ||
          content1.join(' ').localeCompare(content2.join(' ')) ||
          mark1.localeCompare(mark2)
  }

  return logs.sort(comparator)
}

console.log(
  reorderLogFiles(['dig1 8 1 5 1', 'let1 art can', 'dig2 3 6', 'let2 own kit dig', 'let3 art zero'])
)
